### ISSUE TITLE
plotter: plotting an histogram in logscale

### ISSUE THREAD

**Issue by sbinet** [gonum/plot#377]
hi,

consider the following code:

```go
func TestHistogramLogScale(t *testing.T) {
	rnd := rand.New(rand.NewSource(1))
	n := 1000
	vals := make(Values, n)
	for i := 0; i < n; i++ {
		vals[i] = rnd.NormFloat64()
	}

	p, err := plot.New()
	if err != nil {
		log.Panic(err)
	}
	p.Title.Text = "Histogram in logy"
	p.Y.Min = 0.1
	p.Y.Scale = plot.LogScale{}
	p.Y.Tick.Marker = plot.LogTicks{}

	h, err := NewHist(vals, 16)
	if err != nil {
		t.Fatal(err)
	}
	h.Normalize(1)
	p.Add(h)

	err = p.Save(200, 200, "testdata/histogram_log.png")
	if err != nil {
		t.Fatal(err)
	}
}
```

this will panic:
```
=== RUN   TestHistogramLogScale
--- FAIL: TestHistogramLogScale (0.01s)
panic: Values must be greater than 0 for a log scale. [recovered]
	panic: Values must be greater than 0 for a log scale.

goroutine 5 [running]:
testing.tRunner.func1(0xc42005a9c0)
	/usr/lib/go/src/testing/testing.go:622 +0x29d
panic(0x7d7160, 0xc4200f7790)
	/usr/lib/go/src/runtime/panic.go:489 +0x2cf
github.com/gonum/plot.LogTicks.Ticks(0x0, 0x3fda71c2dac35ff2, 0xc420123420, 0x4cbec8, 0x857cfb)
	/home/binet/work/gonum/src/github.com/gonum/plot/axis.go:429 +0x400
github.com/gonum/plot.(*LogTicks).Ticks(0xc89e68, 0x0, 0x3fda71c2dac35ff2, 0x4cc5a5, 0x857cfb, 0x9)
	<autogenerated>:42 +0x5d
github.com/gonum/plot.(*verticalAxis).size(0xc420123af8, 0x40278e0000000000)
	/home/binet/work/gonum/src/github.com/gonum/plot/axis.go:282 +0x7a
github.com/gonum/plot.(*Plot).Draw(0xc4200b3000, 0x7f6fa2e5d428, 0xc420058de0, 0x0, 0x0, 0x4069000000000000, 0x406712b000000000)
	/home/binet/work/gonum/src/github.com/gonum/plot/plot.go:168 +0x232
github.com/gonum/plot.(*Plot).WriterTo(0xc4200b3000, 0x4069000000000000, 0x4069000000000000, 0x86945b, 0x3, 0x0, 0x400aa49aec34e97f, 0x1, 0xc4200f7190)
	/home/binet/work/gonum/src/github.com/gonum/plot/plot.go:455 +0x143
github.com/gonum/plot.(*Plot).Save(0xc4200b3000, 0x4069000000000000, 0x4069000000000000, 0x869444, 0x1a, 0x0, 0x0)
	/home/binet/work/gonum/src/github.com/gonum/plot/plot.go:481 +0x147
github.com/gonum/plot/plotter.TestHistogramLogScale(0xc42005a9c0)
	/home/binet/work/gonum/src/github.com/gonum/plot/plotter/histogram_test.go:129 +0x42b
testing.tRunner(0xc42005a9c0, 0x873468)
	/usr/lib/go/src/testing/testing.go:657 +0x96
created by testing.(*T).Run
	/usr/lib/go/src/testing/testing.go:697 +0x2ca
exit status 2
```

because, even though `p.Y.Min` was set, the empty bins of the histogram will still be "evaluated".

Locally, I have implemented a not completely satisfactory hack: implement "clipping" decorators for `LogScale` and `LogTicks`:

```go
type ClipScale struct {
	Min  float64
	Max  float64
	Norm plot.Normalizer
}

func (cs ClipScale) Normalize(min, max, x float64) float64 {
	min = math.Max(cs.Min, min)
	max = math.Min(cs.Max, max)
	switch {
	case x < cs.Min:
		x = cs.Min
	case x > cs.Max:
		x = cs.Max
	}
	return cs.Norm.Normalize(min, max, x)
}

var _ plot.Normalizer = ClipScale{}

type ClipTicker struct {
	Min    float64
	Max    float64
	Ticker plot.Ticker
}

var _ plot.Ticker = ClipTicker{}

func (ct ClipTicker) Ticks(min, max float64) []plot.Tick {
	min = math.Max(min, ct.Min)
	max = math.Min(max, ct.Max)
	return ct.Ticker.Ticks(min, max)
}
```

and with:
```go
		p.Y.Scale = ClipScale{p.Y.Min, math.Inf(+1), plot.LogScale{}}
		p.Y.Tick.Marker = ClipTicker{p.Y.Min, math.Inf(+1), plot.LogTicks{}}
```

but it's not competely satisfactory because then we get histo-bars for these empty bins:
![histogram_log](https://user-images.githubusercontent.com/1070920/29224890-76d7b4be-7ecc-11e7-9667-d054f35919f1.png)

any idea ?

**Comment by kortschak** [comment]
What do you want the plot to look like?

**Comment by kortschak** [comment]
For ease of debugging: https://play.golang.org/p/ydVV8CRK6Z

**Comment by sbinet** [comment]
well, I'd like the behaviour to be something like that:
 https://github.com/rmadar/ADCTimeSeriePCA/blob/master/Results.png

ie: the empty bins (bottom-right plot) are just not displayed (instead of being drawn as a flat line as in my hacked up version of a "clipper")

(here is the Go equivalent that I had to keep in linear scale: https://github.com/rmadar/ADCTimeSeriePCA/blob/master/go-results.png)

**Comment by sbinet** [comment]
coming back to this (as a colleague of mine also hit that snag...)

I guess we'd need to either:
- introduce the concept of discontiguous domains (so a `plot.Plotter` wouldn't try to display itself on the non-sensical axis domains)
- or introduce some kind of explicit mask that can be applied to `plot.Plotter`

any other idea?

**Comment by ctessum** [comment]
In a histogram, the value that is being represented is supposed to be proportional to the area of the bar. When the base of histogram is not at zero, that is no longer true. So perhaps it is not necessarily a good idea to try to make a histogram with a log axis.

Maybe an alternative solution would be to create a [PDF](http://www.statisticsblog.com/wp-content/uploads/2010/06/Plot_of_Standard_Normal_Density_Function.gif) plotter that would plot a line representing the bin tops, but not plot the bin bottoms?

**Comment by sbinet** [comment]
well, indeed, if you take the definition of a histogram like so, it becomes not as well defined under a log Y-scale...
(although, I suspect you could get away with some carefully chosen renormalization transformation :P)

but in particle physics, we use histograms as a proxy for statistical tools.
I look at histograms as just the number of entries (possibly weighted) that fall within a bunch of bins.
of course, the graphical representation of this binned distribution may vary:
- a rectangular bar
- a marker, possibly decorated with error bars
- a violin-like plot

whatever floats your boat.
but, ultimately, for a bin with 0 entries, there shouldn't be any horizontal bar marking the height of the bar, whatever the Y axis scale.
there aren't any data in that falls inside that bin, so no need to go through the hassle of drawing it and `panic` because of `math.Log(0)`.

**Comment by ctessum** [comment]
That makes sense. However, the current histogram implementation shouldn't create any bins with zero entries: https://github.com/gonum/plot/blob/master/plotter/histogram.go#L154

So, if I understand correctly, the panicking is happening when the plotter tries to draw the bottoms of the rectangles, which are [always at zero](https://github.com/gonum/plot/blob/master/plotter/histogram.go#L81).

I think that immediate problem could be fixed by replacing

``` go
			{trX(bin.Min), trY(0)},
			{trX(bin.Max), trY(0)},
```
with 
``` go
			{trX(bin.Min), trY(p.Y.Min)},
			{trX(bin.Max), trY(p.Y.Min)},
```
in the histogram `Plot` method, and perhaps something similar in the `DataRange` method.

But then we run back into the issue that the apparent relationships in data points are affected by the choice of `p.Y.Min` in a way that is probably not useful.

A more robust solution might be to implement the "marker, possibly decorated with error bars" and/or "violin-like plot" options from above and recommend those for log-transformed plots instead of the rectangle version

**Comment by kortschak** [comment]
ISTM that zero bins should be plotted per @ctessum's claim in https://github.com/gonum/plot/issues/377#issuecomment-394866014, but that there should be a way to not plot a bin. In R, NA is used for this. We have not made a specific magic NaN like this, so we can't use NA, but we could use NaN (or we could decide that a magic NaN is worth including; we have skirted around this idea in the past in floats and I think that it might be a dangerous thing to do).

**Comment by sbinet** [comment]
I went with a slightly different strategy.
I introduced a light wrapper around `plotter.Histogram`: `plotter.LogHistogram` where the minute details of how to cope with y-log scale are localized.

the initial strategy from Chris would have only worked (w/o changing user code) if `plot.DataRanger` were gaining insight of what kind of scale it was applied against...

**Comment by ctessum** [comment]
Could the same goal be achieved by simply adding a "YMin" field to the `Histogram` struct that would specify where the bottom of the histogram bars should be plotted?

**Comment by sbinet** [comment]
I actually started with that (as the commented out hints).
But then the logic was a bit hairy.
It worked fine for linear scale (ie, 0 is fine)

But for the log scale, the user would have to find a good value iteratively.
I like gonum/plot because it creates plots with really sensible defaults... That would break the nice pattern...

**Comment by ctessum** [comment]
I see. I guess the main question then is whether this is a common enough issue to justify the additional API surface here rather than in a separate repository. I've never come across a log-scale histogram before, is it a reasonably common thing?

**Comment by kortschak** [comment]
I'd be happier if this didn't go in this way. I'd much rather see a composable approach. Can a NaN YMin field be used to do this?

**Comment by sbinet** [comment]
Ok, I'll address the initial comments on the PR and create another one
based on a 'YMin' mechanism.

-s

sent from my droid

**Comment by sbinet** [comment]
it's there: https://github.com/gonum/plot/pull/453

but... the `YMin` mechanism is brittle.
in both #452 and #453, if we were to plot 2 histograms on the same plot, the probability that the lower ymin value `DataRange()` would come up with to be the same for the 2 histos is pretty low.

I am wondering whether there isn't a need really to inject some knowledge about the overall plot layout...
otherwise, you'd get something like that:
![histogram_logy](https://user-images.githubusercontent.com/1070920/41191727-fc2e9082-6bf3-11e8-8a5d-ac36b55ce27e.png)

**Comment by sbinet** [comment]
> I see. I guess the main question then is whether this is a common enough issue to justify the additional API surface here rather than in a separate repository. I've never come across a log-scale histogram before, is it a reasonably common thing?

@ctessum, well it's reasonably common in particle physics. one example that comes to mind is when one wants to compare physics background simulations of 2 MonteCarlo generators (or MC(signal+bkg) and data), where backgrounds processes usually present an exponantial behaviour.
_e.g._ https://physicsforme.com/2012/05/31/quark-excitement-is-there-anything-smaller/
![hist-example](https://physicsforme.files.wordpress.com/2012/05/atlas1.jpg)

**Comment by aphecetche** [comment]
Dear all,

Was there a conclusion here ? 
Got burnt by the panic today trying to plot a simple histogram in log scale :-(

**Comment by kortschak** [comment]
@sbinet That plot (the upper panel) looks like a stepped wiggle plot rather than a histogram. That class of plot certainly warrant a log scale.

Is there a reason not to use NaN here?

**Comment by sbinet** [comment]
@aphecetche this kind of died out (my `ClipScale` hack worked for my colleague at LPC.)

@kortschak I guess we need a common vocabulary for "histogram" :)
the physics process being studied in this particular plot has some background noise (ie: not interesting data) that has an exp behaviour. In this particular case there's no empty bin (except perhaps at high mass where it's expensive to generate/simulate MonteCarlo data) but it's relatively common to get such empty bins.
I am unsure about using a NaN.
could we use "quite NaNs" for masking these empty bins? (to make sure we don't sweep lurking issues under the rug, and properly panic on NaNs that are coming from the analysis pipeline.)

**Comment by kortschak** [comment]
I don't understand the significance of quiet NaN here. All NaN are quiet in Go, and `plotter.Histogram` does not do any explicit NaN check. I was thinking of just omitting bins that have a NaN; we should note this on the histgram docs. The behaviour would be to not include those NaN values in the generation of ticks and extension of the axis range.

I still don't see a valid reason for including a 0 on a log scale. For example, the figure above in https://github.com/gonum/plot/issues/377#issuecomment-395967957 doesn't include a zero on a log scale, there is a zero there, but that is not a log scale.

**Comment by sbinet** [comment]
ok. new year, new plan, new fix, new PR: #493.
PTAL.
